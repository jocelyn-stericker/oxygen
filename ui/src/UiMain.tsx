import { RenderMode, UiState } from "oxygen-core";
import React, { useState, useCallback, useRef, useReducer } from "react";
import cx from "classnames";

import Toaster, { ToasterInterface } from "./toaster";
import ClipList from "./ClipList";
import RecordTab from "./RecordTab";
import CurrentClip from "./CurrentClip";

function nativeLog(level: string, log: string) {
  if (level === "error") {
    console.error(log);
  } else if (level === "trace") {
    console.trace(log);
  } else if (level === "warn") {
    console.warn(log);
  } else if (level === "info") {
    console.info(log);
  } else if (level === "debug") {
    console.debug(log);
  } else {
    console.log(log, `(Note: unknown log level ${level}`);
  }
}

export default function Main({ inMemory }: { inMemory: boolean }) {
  // Hack to force a re-render when the state changes.
  const [, forceUpdate] = useReducer(() => ({}), []);
  const [uiState] = useState(
    () => new UiState(forceUpdate, nativeLog, inMemory)
  );
  const toaster = useRef<ToasterInterface>(null);

  const drawCurrentClip = useCallback(
    (width: number, height: number) => uiState.drawCurrentClip(width, height),
    [uiState]
  );

  const transcribe = useCallback(() => uiState.transcribe(), [uiState]);

  const handlePlay = useCallback(
    (cb: () => void) => {
      uiState.play(cb);
    },
    [uiState]
  );

  const handleStop = useCallback(() => uiState.stop(), [uiState]);

  const handleSeek = useCallback(
    (timePercent: number) => {
      uiState.seek(timePercent);
    },
    [uiState]
  );

  const handleRename = useCallback(
    (name: string) => {
      try {
        const prevName = uiState.currentClip.name;
        if (name !== prevName) {
          uiState.renameCurrentClip(name);
          toaster.current.info(`Renamed "${prevName}" to "${name}"`);
        }
      } catch (err) {
        if (err instanceof Error) {
          // TODO: stable interface for error messages and/or tests
          if (err.message == "UNIQUE constraint failed: clips.name") {
            toaster.current.error("This name is taken by another clip.");
          } else {
            toaster.current.error(
              "Something went wrong when renaming this clip."
            );
          }
        }
      }
    },
    [uiState]
  );

  const handleDelete = useCallback(() => {
    const prevName = uiState.currentClip.name;
    uiState.deleteCurrentClip();
    toaster.current.info(
      `Deleted "${prevName}"`,
      {
        text: "Undo",
        cb: () => {
          uiState.undeleteCurrentClip();
        },
      },
      "undoDeleteCurrentClip"
    );
  }, [uiState]);

  const handleSetRenderMode = useCallback(
    (renderMode: RenderMode) => {
      uiState.setRenderMode(renderMode);
    },
    [uiState]
  );

  const handleRecord = useCallback(() => {
    uiState.record();
  }, [uiState]);

  const handleSetTabRecord = useCallback(() => {
    uiState.setCurrentTabRecord();
  }, [uiState]);

  const handleSetCurrentClipId = useCallback(
    (clipId: number) => {
      uiState.setCurrentClipId(clipId);
    },
    [uiState]
  );

  const handleExport = useCallback(
    (clipId: number) => {
      return uiState.export(clipId);
    },
    [uiState]
  );

  const [dragOver, setDragOver] = useState<boolean | "invalid">(false);

  return (
    <div
      className="w-full h-full flex flex-row"
      onDragOver={(ev) => {
        ev.preventDefault();
        if (
          [...ev.dataTransfer.items].filter(
            (item) => item.type === "audio/wav" || item.type === "audio/mpeg"
          ).length > 0
        ) {
          setDragOver(true);
        } else {
          setDragOver("invalid");
        }
      }}
      onDragLeave={(ev) => {
        ev.preventDefault();
        setDragOver(false);
      }}
      onDrop={(ev) => {
        for (const item of ev.dataTransfer.items) {
          if (item.type === "audio/wav" || item.type === "audio/mpeg") {
            try {
              uiState.import(item.getAsFile().path);
              toaster.current.info(`Imported ${item.getAsFile().name}.`);
            } catch (err) {
              toaster.current.error(
                `Could not import ${item.getAsFile().name}: ${err.toString()}`
              );
            }
          } else {
            toaster.current.error(
              `Count not import ${
                item.getAsFile().name
              } because the file type is unsupported.`
            );
          }
        }
        setDragOver(false);
      }}
    >
      <ClipList
        clips={uiState.getClips().reverse()}
        recordTabSelected={uiState.recordTabSelected}
        currentClipId={uiState.currentClipId}
        onSetCurrentTabRecord={handleSetTabRecord}
        onSetCurrentClipId={handleSetCurrentClipId}
        onExport={handleExport}
      />
      {uiState.currentClipId != null && (
        <CurrentClip
          clip={uiState.currentClip}
          drawCurrentClip={drawCurrentClip}
          transcribe={transcribe}
          time={uiState.time}
          streaming={uiState.streaming}
          renderMode={uiState.renderMode}
          onPlay={handlePlay}
          onStop={handleStop}
          onSeek={handleSeek}
          onRename={handleRename}
          onDelete={handleDelete}
          onSetRenderMode={handleSetRenderMode}
          timeStart={Number(uiState.timeStart)}
          timeEnd={Number(uiState.timeEnd)}
        />
      )}
      {uiState.recordTabSelected && (
        <RecordTab
          streaming={uiState.streaming}
          renderMode={uiState.renderMode}
          drawCurrentClip={drawCurrentClip}
          onRecord={handleRecord}
          onStop={handleStop}
          onSetRenderMode={handleSetRenderMode}
          timeStart={Number(uiState.timeStart)}
          timeEnd={Number(uiState.timeEnd)}
        />
      )}
      {dragOver && (
        <div className="absolute h-full w-full p-2">
          <div
            className={cx(
              "relative h-full w-full border-4 text-2xl flex items-center font-bold justify-center",
              dragOver === "invalid" && "border-red-600/75  bg-red-200/75",
              dragOver === true && "border-blue-600/75  bg-blue-200/75"
            )}
          >
            {dragOver === true && <div>Drop audio clips here.</div>}
            {dragOver === "invalid" && (
              <div>This file type is not supported.</div>
            )}
          </div>
        </div>
      )}
      <Toaster ref={toaster} />
    </div>
  );
}
