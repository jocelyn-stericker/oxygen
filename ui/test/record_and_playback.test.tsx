import React from "react";
import { act, fireEvent, render, within } from "@testing-library/react";
import diff from "snapshot-diff";
import postcss from "postcss";
import tailwind from "tailwindcss";

import UiMain from "../src/UiMain";

beforeAll(async () => {
  jest
    .spyOn(Date.prototype, "toLocaleDateString")
    .mockReturnValue("Mocked Date");
  jest
    .spyOn(Date.prototype, "toLocaleTimeString")
    .mockReturnValue("Mocked Date");

  const styleSheet = document.createElement("style");
  styleSheet.innerText = (
    await postcss([tailwind("tailwind.config.js")]).process(
      " @tailwind base; @tailwind components; @tailwind utilities;",
      {
        from: "../src/index.css",
      },
    )
  ).css;
  document.head.appendChild(styleSheet);
});

describe("app [integration]", () => {
  it("can record and playback a clip", async () => {
    const app = render(
      <div style={{ width: 1024, height: 768 }}>
        <UiMain inMemory={true} />
      </div>,
    );
    const startRecording = app.getByRole("button", {
      name: "Start Recording",
    });
    fireEvent.click(startRecording);

    let nextFragment = app.asFragment();
    const initialFragment = nextFragment;
    expect(nextFragment).toMatchSnapshot("1 initial");
    let prevFragment = nextFragment;

    const stopRecording = await app.findByRole("button", {
      name: "Complete Recording",
    });
    nextFragment = app.asFragment();
    expect(diff(prevFragment, nextFragment)).toMatchSnapshot("1 recording");
    prevFragment = nextFragment;

    await act(async () => {
      await new Promise((res) => setTimeout(res, 200));
      fireEvent.click(stopRecording);
      await new Promise((res) => setTimeout(res, 200));
    });

    const play = await app.findByRole("button", { name: "Play" });
    // Rename the clip since it has a date in it.
    const clipName = app.getByTestId("current-clip-name") as HTMLInputElement;
    expect(clipName.value).toMatch(/20\d\d-\d\d-\d\d .*/);
    fireEvent.focus(clipName);
    fireEvent.change(clipName, { target: { value: "New clip name" } });
    fireEvent.blur(clipName);

    const dismissRename = within(
      app.getByText(/Renamed.*to.*/).parentElement,
    ).getByRole("button", { name: "Dismiss" });
    fireEvent.click(dismissRename);

    await app.findByDisplayValue("New clip name");

    nextFragment = app.asFragment();
    expect(diff(prevFragment, nextFragment)).toMatchSnapshot(
      "2 done recording",
    );
    prevFragment = nextFragment;

    fireEvent.click(play);

    const pause = await app.findByRole("button", { name: "Pause" });
    await app.findByRole("button", { name: "Pause" });

    fireEvent.click(pause);

    fireEvent.click(
      app.getByRole("tab", { name: "Record New Clip", selected: false }),
    );
    await app.findByRole("tab", { name: "Record New Clip", selected: true });

    nextFragment = app.asFragment();
    expect(diff(initialFragment, nextFragment)).toMatchSnapshot(
      "3 delta from initial",
    );
  });
});
