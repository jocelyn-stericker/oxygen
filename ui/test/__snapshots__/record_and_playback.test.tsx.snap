// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app [integration] can record and playback a clip: 1 initial 1`] = `
<DocumentFragment>
  <div
    style="width: 1024px; height: 768px;"
  >
    <div
      class="w-full h-full flex flex-row"
    >
      <div
        class="w-72 border-r-purple-900 border-r-2 h-full divide-y divide-purple-200 overflow-y-auto relative"
      >
        <div
          class="m-2"
        >
          <input
            class="self-center p-2 text-m font-bold overflow-ellipses overflow-hidden border-2 border-purple-200 rounded-md focus:border-purple-900 outline-purple-900 text-purple-900 flex-grow transition-all w-full"
            data-testid="filter-clip"
            placeholder="Search clips"
            title="Search clips"
            value=""
          />
        </div>
        <ul
          role="tablist"
        >
          <li
            aria-selected="true"
            class="hover:bg-purple-100 cursor-pointer text-purple-900 overflow-hidden bg-purple-900 text-white hover:bg-purple-900 cursor-default"
            data-testid="record-item"
            role="tab"
          >
            <h2
              class="p-4 text-m font-bold overflow-ellipsis overflow-hidden flex flex-row justify-center"
            >
              <svg
                class="h-6 w-6"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M19 11a7 7 0 01-7 7m0 0a7 7 0 01-7-7m7 7v4m0 0H8m4 0h4m-4-8a3 3 0 01-3-3V5a3 3 0 116 0v6a3 3 0 01-3 3z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
               Record New Clip
            </h2>
          </li>
          <div
            class="text-center text-gray-500 italic p-2"
            data-testid="cliplist-placeholder"
          >
            Your clips will appear here.
          </div>
        </ul>
      </div>
      <div
        class="flex flex-col flex-grow"
      >
        <div
          class="flex-grow relative overflow-hidden"
        >
          <canvas
            class="absolute w-full h-full"
            data-testid="current-clip-view"
            height="688"
            style="width: 736px; height: 688px; transform: scale(1); transform-origin: left top;"
            width="736"
          />
          <div
            class="absolute w-[1px] bg-blue-400 h-full"
            data-testid="current-clip-cursor"
            style="left: 100%;"
          />
        </div>
        <div
          class="flex flex-row mb-4"
        >
          <div
            class="flex-grow"
          />
          <button
            class="px-4 rounded-md m-auto text-lg flex border-2 h-16 items-center bg-purple-900 text-white hover:bg-purple-800"
            data-testid="toggle-record"
          >
            <svg
              class="h-6 w-6"
              fill="none"
              stroke="currentColor"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M19 11a7 7 0 01-7 7m0 0a7 7 0 01-7-7m7 7v4m0 0H8m4 0h4m-4-8a3 3 0 01-3-3V5a3 3 0 116 0v6a3 3 0 01-3 3z"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
              />
            </svg>
            <span
              class="w-2"
            />
            Start Recording
          </button>
          <div
            class="flex-grow"
          />
        </div>
      </div>
      <div
        class="fixed left-1/2 bottom-0 flex flex-col-reverse"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`app [integration] can record and playback a clip: 1 recording 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -80,31 +80,37 @@
          >
            <div
              class="flex-grow"
            />
            <button
-             class="px-4 rounded-md m-auto text-lg flex border-2 h-16 items-center bg-purple-900 text-white hover:bg-purple-800"
+             class="px-4 rounded-md m-auto text-lg flex border-2 h-16 items-center bg-white border-purple-900 text-purple-900 hover:bg-purple-100"
              data-testid="toggle-record"
            >
              <svg
                class="h-6 w-6"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
-                 d="M19 11a7 7 0 01-7 7m0 0a7 7 0 01-7-7m7 7v4m0 0H8m4 0h4m-4-8a3 3 0 01-3-3V5a3 3 0 116 0v6a3 3 0 01-3 3z"
+                 d="M21 12a9 9 0 11-18 0 9 9 0 0118 0z"
+                 stroke-linecap="round"
+                 stroke-linejoin="round"
+                 stroke-width="2"
+               />
+               <path
+                 d="M9 10a1 1 0 011-1h4a1 1 0 011 1v4a1 1 0 01-1 1h-4a1 1 0 01-1-1v-4z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
              <span
                class="w-2"
              />
-             Start Recording
+             Complete Recording
            </button>
            <div
              class="flex-grow"
            />
          </div>"
`;

exports[`app [integration] can record and playback a clip: 2 done recording 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -21,12 +21,12 @@
          </div>
          <ul
            role="tablist"
          >
            <li
-             aria-selected="true"
-             class="hover:bg-purple-100 cursor-pointer text-purple-900 overflow-hidden bg-purple-900 text-white hover:bg-purple-900 cursor-default"
+             aria-selected="false"
+             class="hover:bg-purple-100 cursor-pointer text-purple-900 overflow-hidden"
              data-testid="record-item"
              role="tab"
            >
              <h2
                class="p-4 text-m font-bold overflow-ellipsis overflow-hidden flex flex-row justify-center"
@@ -46,74 +46,128 @@
                  />
                </svg>
                 Record New Clip
              </h2>
            </li>
+           <li
+             aria-selected="true"
+             class="p-2 hover:bg-purple-100 cursor-pointer text-purple-900 overflow-hidden bg-purple-900 text-white hover:bg-purple-900 cursor-default"
+             data-testid="clip-1"
+             role="tab"
+           >
+             <h2
+               class="text-m font-bold overflow-ellipsis overflow-hidden"
+               title="New clip name"
+             >
+               New clip name
+             </h2>
              <div
-             class="text-center text-gray-500 italic p-2"
-             data-testid="cliplist-placeholder"
+               class="flex flex-row"
              >
-             Your clips will appear here.
+               <div
+                 class="text-xs font-light"
+               >
+                 Mocked Date at Mocked Date
+               </div>
              </div>
+           </li>
          </ul>
        </div>
        <div
          class="flex flex-col flex-grow"
+       >
+         <div
+           class="flex flex-row"
+         >
+           <input
+             class="self-center p-2 m-2 text-m font-bold overflow-ellipses overflow-hidden border-2 border-purple-200 rounded-md focus:border-purple-900 outline-purple-900 text-purple-900 flex-grow transition-all"
+             data-testid="current-clip-name"
+             value="New clip name"
+           />
+           <button
+             class="p-2 m-2 ml-0 text-purple-900 cursor-pointer border-2 border-transparent hover:border-red-900 rounded-full hover:bg-red-100 hover:text-red-900"
+             data-testid="current-clip-delete"
+             title="Delete this clip"
+           >
+             <svg
+               class="h-6 w-6"
+               fill="none"
+               stroke="currentColor"
+               stroke-width="2"
+               viewBox="0 0 24 24"
+               xmlns="http://www.w3.org/2000/svg"
              >
+               <path
+                 d="M19 7l-.867 12.142A2 2 0 0116.138 21H7.862a2 2 0 01-1.995-1.858L5 7m5 4v6m4-6v6m1-10V4a1 1 0 00-1-1h-4a1 1 0 00-1 1v3M4 7h16"
+                 stroke-linecap="round"
+                 stroke-linejoin="round"
+               />
+             </svg>
+           </button>
+         </div>
          <div
            class="flex-grow relative overflow-hidden"
          >
            <canvas
              class="absolute w-full h-full"
              data-testid="current-clip-view"
-             height="688"
-             style="width: 736px; height: 688px; transform: scale(1); transform-origin: left top;"
+             height="628"
+             style="width: 736px; height: 628px; transform: scale(1); transform-origin: left top;"
              width="736"
            />
            <div
              class="absolute w-[1px] bg-blue-400 h-full"
              data-testid="current-clip-cursor"
-             style="left: 100%;"
+             style="left: 0%;"
            />
          </div>
          <div
            class="flex flex-row mb-4"
          >
            <div
+             class="flex self-center font-mono text-purple-900 mx-2 w-20"
+             data-testid="current-clip-time"
+           >
+             000.00
+           </div>
+           <div
              class="flex-grow"
            />
            <button
-             class="px-4 rounded-md m-auto text-lg flex border-2 h-16 items-center bg-white border-purple-900 text-purple-900 hover:bg-purple-100"
-             data-testid="toggle-record"
+             class="px-4 rounded-md m-auto text-lg flex border-2 h-16 items-center bg-purple-900 text-white hover:bg-purple-800"
+             data-testid="current-clip-toggle-playback"
            >
              <svg
                class="h-6 w-6"
                fill="none"
                stroke="currentColor"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
-                 d="M21 12a9 9 0 11-18 0 9 9 0 0118 0z"
+                 d="M14.752 11.168l-3.197-2.132A1 1 0 0010 9.87v4.263a1 1 0 001.555.832l3.197-2.132a1 1 0 000-1.664z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
                <path
-                 d="M9 10a1 1 0 011-1h4a1 1 0 011 1v4a1 1 0 01-1 1h-4a1 1 0 01-1-1v-4z"
+                 d="M21 12a9 9 0 11-18 0 9 9 0 0118 0z"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                />
              </svg>
              <span
                class="w-2"
              />
-             Complete Recording
+             Play
            </button>
            <div
              class="flex-grow"
+           />
+           <div
+             class="w-20 mx-2"
            />
          </div>
        </div>
        <div
          class="fixed left-1/2 bottom-0 flex flex-col-reverse""
`;

exports[`app [integration] can record and playback a clip: 3 delta from initial 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -46,16 +46,32 @@
                  />
                </svg>
                 Record New Clip
              </h2>
            </li>
+           <li
+             aria-selected="false"
+             class="p-2 hover:bg-purple-100 cursor-pointer text-purple-900 overflow-hidden"
+             data-testid="clip-1"
+             role="tab"
+           >
+             <h2
+               class="text-m font-bold overflow-ellipsis overflow-hidden"
+               title="New clip name"
+             >
+               New clip name
+             </h2>
              <div
-             class="text-center text-gray-500 italic p-2"
-             data-testid="cliplist-placeholder"
+               class="flex flex-row"
              >
-             Your clips will appear here.
+               <div
+                 class="text-xs font-light"
+               >
+                 Mocked Date at Mocked Date
+               </div>
              </div>
+           </li>
          </ul>
        </div>
        <div
          class="flex flex-col flex-grow"
        >"
`;
